content/chapters/compute/lab/support/user-level-threads/threads_and_fibers.cc

```diff
@@ -0,0 +1,90 @@
+#include <vector>
+#include <thread>
+
+#include <boost/fiber/all.hpp>
+#include <boost/fiber/detail/thread_barrier.hpp>
+
+#define NUM_FIBERS 10
+#define NUM_THREADS 3
+
+static boost::fibers::fiber_specific_ptr<int> fiber_id;
+static thread_local int thread_id;
+
+static std::mutex print_mutex;
+static std::size_t fiber_count{ 0 };
+static std::mutex count_mutex{};
+typedef std::unique_lock<std::mutex> lock_type;
+static boost::fibers::condition_variable_any cnd_count{};
+
+static boost::fibers::detail::thread_barrier thread_barrier(NUM_THREADS);
+
+void fiber_work(int id)
+{
+	fiber_id.reset(new int(id));
+
+	for (int i = 0; i < 10; i++)
+	{
+		print_mutex.lock();
+		std::cout << "Fiber " << *fiber_id << ", on thread " << thread_id << ", yields" << std::endl;
+		print_mutex.unlock();
+
+		boost::this_fiber::yield();
+	}
+	lock_type lk(count_mutex);
+	if (0 == --fiber_count) {
+		lk.unlock();
+		cnd_count.notify_all();
+	}
+	// lk is unlocked here
+}
+
+void thread_work(int id)
+{
+	thread_id = id;
+
+	print_mutex.lock();
+	std::cout << "thread " << thread_id << " started" << std::endl;
+	print_mutex.unlock();
+
+	// TODO: use the `work_stealing` scheduler
+	boost::fibers::use_scheduling_algorithm<boost::fibers::algo::shared_work>();
+
+	thread_barrier.wait();
+
+	{
+		lock_type lk(count_mutex);
+		cnd_count.wait( lk, [](){ return 0 == fiber_count; } );
+	}
+	// lk is unlocked here
```

Comment:
Ditto: unlock `lk` manually. And why do you create another scope here?

Reply:
when `unique_lock` gets out of scope, it is unlocked automatically